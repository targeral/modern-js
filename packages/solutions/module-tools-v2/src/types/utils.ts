export type DeepPartial<T> = keyof T extends never
  ? T
  : { [k in keyof T]?: DeepPartial<T[k]> };
