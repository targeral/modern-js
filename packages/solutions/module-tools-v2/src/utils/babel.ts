import type { AliasOption, BundlelessGeneratorDtsConfig } from '../types';

export const getFinalAlias: any = async (
  alias: AliasOption,
  config: BundlelessGeneratorDtsConfig,
) => {
  const { getAlias } = await import('@modern-js/utils');
  const aliasConfig = getAlias(alias as any, {
    appDirectory: config.appDirectory,
    tsconfigPath: config.tsconfigPath,
  });
  // 排除内部别名，因为不需要处理
  const finalPaths: Record<string, string | string[]> = {};
  const internalAliasPrefix = '@modern-js/runtime';
  if (aliasConfig.paths && typeof aliasConfig.paths === 'object') {
    const pathsKey = Object.keys(aliasConfig.paths);
    for (const key of pathsKey) {
      if (!key.includes(internalAliasPrefix)) {
        finalPaths[key] = aliasConfig.paths[key];
      }
    }
  }

  aliasConfig.paths = finalPaths;
  return aliasConfig;
};
