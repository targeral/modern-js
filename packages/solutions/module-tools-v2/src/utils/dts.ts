import path from 'path';
import t from '@babel/types';
import type * as tt from '@babel/types';
import type { NodePath } from '@babel/traverse';
import type {
  ITsconfig,
  BundlelessGeneratorDtsConfig,
  BuildCommandOptions,
  BaseBuildConfig,
  AliasOption,
} from '../types';

export const generatorTsConfig = async (
  projectTsconfig: ITsconfig,
  config: BundlelessGeneratorDtsConfig,
) => {
  const { nanoid, fs, lodash } = await import('@modern-js/utils');

  const { appDirectory, sourceDir, distAbsPath, tsconfigPath } = config;
  const tempPath = path.resolve(appDirectory, './node_modules');
  const resolvePath = path.relative(tempPath, sourceDir);

  const rootDir = resolvePath;
  const baseUrl = projectTsconfig.compilerOptions?.baseUrl
    ? path.join(appDirectory, projectTsconfig.compilerOptions?.baseUrl)
    : appDirectory;
  // if include = ['src'], final include should be ['../src']
  const include = [resolvePath];

  const resetConfig: ITsconfig = {
    compilerOptions: {
      rootDir,
      baseUrl,
      // Ensure that .d.ts files are created by tsc, but not .js files
      declaration: true,
      emitDeclarationOnly: true,
      outDir: distAbsPath,
    },
    include,
    exclude: projectTsconfig.exclude ?? [],
  };

  // extends: '../tsconfig.json'
  if (projectTsconfig.extends) {
    resetConfig.extends = projectTsconfig.extends.startsWith('.')
      ? path.join(
          path.relative(`${tempPath}/tsconfig.json`, tsconfigPath),
          projectTsconfig.extends,
        )
      : projectTsconfig.extends;
  }

  const recommendOption = {
    // Ensure that Babel can safely transpile files in the TypeScript project
    compilerOptions: {
      isolatedModules: true,
    },
  };

  const tempTsconfigPath = path.join(
    tempPath,
    `tsconfig.${Date.now()}.${nanoid()}.json`,
  );
  fs.ensureFileSync(tempTsconfigPath);

  const deepMerge = lodash.merge;
  fs.writeJSONSync(
    tempTsconfigPath,
    deepMerge(
      recommendOption,
      projectTsconfig,
      // 此处是必须要覆盖用户默认配置
      resetConfig,
    ),
  );

  return tempTsconfigPath;
};

export const getTscBinPath = async (appDirectory: string) => {
  const { fs } = await import('@modern-js/utils');
  const tscBinFile = path.join(appDirectory, './node_modules/.bin/tsc');

  if (!fs.existsSync(tscBinFile)) {
    throw new Error(
      'Failed to excute the `tsc` command, please check if `typescript` is installed correctly in the current directory.',
    );
  }

  return tscBinFile;
};

export const resolveAlias = async (
  config: BundlelessGeneratorDtsConfig,
  watchFilenames: string[] = [],
) => {
  const { globby, fs } = await import('@modern-js/utils');
  const { getFinalAlias } = await import('./babel');
  const { transformDtsAlias } = await import('./tspaths-transform');
  const { distAbsPath, alias } = config;
  const defaultPaths = { '@': ['./src'] };
  const dtsDistPath = `${distAbsPath}/**/*.d.ts`;
  const dtsFilenames =
    watchFilenames.length > 0
      ? watchFilenames
      : globby.sync(dtsDistPath, { absolute: true });
  const finalAlias = await getFinalAlias(alias, config);
  const mergedPaths = finalAlias.isTsPath
    ? finalAlias.paths || {}
    : { ...defaultPaths, ...(finalAlias.paths || {}) };
  const result = transformDtsAlias({
    filenames: dtsFilenames,
    baseUrl: distAbsPath,
    paths: mergedPaths,
  });
  for (const r of result) {
    fs.writeFileSync(r.path, r.content);
  }
};

export const matchesPattern = (calleePath: NodePath, pattern: string) => {
  const { node } = calleePath;

  if (t.isMemberExpression(node)) {
    return calleePath.matchesPattern(pattern);
  }

  if (!t.isIdentifier(node) || pattern.includes('.')) {
    return false;
  }

  const name = pattern.split('.')[0];

  return node.name === name;
};

export const isImportCall = (calleePath: NodePath<tt.CallExpression>) => {
  return t.isImport(calleePath.node.callee);
};

export const verifyTsConfigPaths = async (
  tsconfigAbsPath: string,
  userAliases?: AliasOption,
) => {
  const { readTsConfigByFile, chalk } = await import('@modern-js/utils');
  if (!userAliases) {
    return;
  }

  const paths = Object.keys(
    readTsConfigByFile(tsconfigAbsPath).compilerOptions?.paths || {},
  ).map(key => key.replace(/\/\*$/, ''));

  Object.keys(userAliases).forEach(name => {
    if (paths.includes(name)) {
      throw new Error(
        chalk.red(
          `It looks like you have configured the alias ${chalk.bold(
            name,
          )} in both the modern.config file and tsconfig.json.\n Please remove the configuration in modern.config file and just keep the configuration in tsconfig.json.`,
        ),
      );
    }
  });
};

export const assignTsConfigPath = async (
  config: BaseBuildConfig,
  options: BuildCommandOptions,
) => {
  const { defaultTsConfigPath } = await import('../constants/dts');

  // user run `build --tsconfig './tsconfig.build.json'`
  if (
    typeof options.tsconfig === 'string' &&
    options.tsconfig !== defaultTsConfigPath
  ) {
    config.dts = {
      only: false,
      distPath: './',
      ...(config.dts ?? {}),
      tsconfigPath: options.tsconfig,
    };
  }

  return config;
};
