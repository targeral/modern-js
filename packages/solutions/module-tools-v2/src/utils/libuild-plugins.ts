import type { LibuildPlugin } from '@modern-js/libuild';
import type { BaseBuildConfig } from '../types';

export const watchPlugin = (config: BaseBuildConfig): LibuildPlugin => {
  return {
    name: 'watch-plugin',
    apply(compiler) {
      compiler.hooks.watchChange.tap('watch-plugin', async () => {
        const { watchSectionTitle } = await import('./log');
        const { SectionTitleStatus } = await import('../constants/log');
        const titleText = `[${
          config.buildType === 'bundle' ? 'Bundle' : 'Bundleless'
        }:${config.format}_${config.target}]`;
        console.info(
          await watchSectionTitle(titleText, SectionTitleStatus.Log),
        );
      });
    },
  };
};
