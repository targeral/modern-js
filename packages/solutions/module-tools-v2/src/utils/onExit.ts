export const addExitListener = async (fn: () => Promise<void>) => {
  const { default: onExit } = await import('signal-exit');

  onExit(async code => {
    if (code === 0) {
      return;
    }

    await fn();
  });
};
