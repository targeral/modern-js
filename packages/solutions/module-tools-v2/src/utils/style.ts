import type { AcceptedPlugin } from 'postcss';
import type { PluginAPI } from '@modern-js/core';
import type { PostcssOptions } from '@modern-js/libuild';
import type {
  UserConfig,
  PostCSSConfigUtils,
  LessOptions,
  SassOptions,
} from '../types';
import type { ModuleToolsHooks } from '../types/hooks';

export const getLessConfig = async (config: UserConfig) => {
  const { applyOptionsChain } = await import('@modern-js/utils');
  const mergedOptions = applyOptionsChain<LessOptions, never>(
    { javascriptEnabled: true },
    config?.tools?.less || {},
  );

  return mergedOptions;
};

export const getSassConfig = async (config: UserConfig) => {
  const { applyOptionsChain } = await import('@modern-js/utils');
  const mergedOptions = applyOptionsChain<SassOptions, never>(
    {},
    config.tools?.sass || {},
  );

  return mergedOptions;
};

export const getPostcssConfig = async (config: UserConfig) => {
  const { applyOptionsChain } = await import('@modern-js/utils');
  const { getCompiledPath } = await import('./path');
  const extraPlugins: AcceptedPlugin[] = [];

  const utils = {
    addPlugins(plugins: AcceptedPlugin | AcceptedPlugin[]) {
      if (Array.isArray(plugins)) {
        extraPlugins.push(...plugins);
      } else {
        extraPlugins.push(plugins);
      }
    },
  };

  const mergedConfig = applyOptionsChain<PostcssOptions, PostCSSConfigUtils>(
    {
      plugins: [
        require(getCompiledPath('postcss-flexbugs-fixes')),
        require(getCompiledPath('postcss-custom-properties')),
        require(getCompiledPath('postcss-initial')),
        require(getCompiledPath('postcss-page-break')),
        require(getCompiledPath('postcss-font-variant')),
        require(getCompiledPath('postcss-media-minmax')),
        require(getCompiledPath('postcss-nesting')),
      ].filter(Boolean),
    },
    (config?.tools?.postcss as any) || {},
    utils,
  );
  if (extraPlugins.length) {
    mergedConfig.plugins!.push(...extraPlugins);
  }

  return mergedConfig;
};

export const getStyleConfig = async (api: PluginAPI<ModuleToolsHooks>) => {
  const config = api.useResolvedConfigContext() as unknown as UserConfig;
  const postcssConfig = await getPostcssConfig(config);
  const lessConfig = await getLessConfig(config);
  const sassConfig = await getSassConfig(config);

  const runner = api.useHookRunners();
  const tailwindPlugin = await runner.addTailwindCssConfig(undefined, {
    onLast: async (_: any) => undefined,
  });
  if (tailwindPlugin) {
    postcssConfig.plugins?.push(tailwindPlugin);
  }

  return {
    less: lessConfig,
    sass: sassConfig,
    postcss: postcssConfig,
  };
};
