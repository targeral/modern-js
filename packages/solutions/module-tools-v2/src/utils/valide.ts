import path from 'path';
import { validAlias } from '@modern-js/utils';
import type { NormalizedConfig, PluginAPI } from '@modern-js/core';
import type { ModuleToolsHooks } from '../types/hooks';
import type { BuildCommandOptions } from '../types/command';

export interface IValideOption {
  modernConfig: NormalizedConfig;
  tsconfigPath: string;
}

export const valideBeforeTask = (
  api: PluginAPI<ModuleToolsHooks>,
  options: BuildCommandOptions,
) => {
  const modernConfig = api.useResolvedConfigContext();
  const { appDirectory } = api.useAppContext();
  const tsconfigPath = path.join(
    appDirectory,
    options.tsconfig ?? 'tsconfig.json',
  );
  const modernConfigValidResult = modernConfigValid(modernConfig, {
    tsconfigPath,
  });
  if (modernConfigValidResult) {
    console.error(modernConfigValidResult);
    // eslint-disable-next-line no-process-exit
    process.exit(0);
  }
};

export const modernConfigValid = (
  modernConfig: NormalizedConfig,
  option: { tsconfigPath: string },
) => {
  const valids = [validAlias];

  for (const validFn of valids) {
    const result = validFn(modernConfig, option);
    if (result) {
      return result;
    }
  }

  return null;
};
