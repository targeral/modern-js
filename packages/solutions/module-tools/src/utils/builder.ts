import path from 'path';
import { logger } from '@modern-js/utils/logger';
import type { BaseBuildConfig, ExternalHelpers } from '../types/config';

export const getFinalExternals = async (
  config: BaseBuildConfig,
  options: { appDirectory: string },
) => {
  const { autoExternal, externals } = config;
  const { appDirectory } = options;

  if (typeof autoExternal === 'boolean') {
    if (!autoExternal) {
      return externals || [];
    }

    const deps = await getAllDeps(appDirectory, {
      dependencies: true,
      peerDependencies: true,
    });
    return [
      ...deps.map(dep => new RegExp(`^${dep}($|\\/|\\\\)`)),
      ...(externals || []),
    ];
  }

  const deps = await getAllDeps(appDirectory, autoExternal);
  return [
    ...deps.map(dep => new RegExp(`^${dep}($|\\/|\\\\)`)),
    ...(externals || []),
  ];
};

export const getAllDeps = async <T>(
  appDirectory: string,
  options: {
    dependencies?: boolean;
    devDependencies?: boolean;
    peerDependencies?: boolean;
  } = {},
) => {
  const { fs } = await import('@modern-js/utils');
  try {
    const json = JSON.parse(
      fs.readFileSync(path.resolve(appDirectory, './package.json'), 'utf8'),
    );

    let deps: string[] = [];

    if (options.dependencies) {
      deps = [
        ...deps,
        ...Object.keys((json.dependencies as T | undefined) || {}),
      ];
    }

    if (options.devDependencies) {
      deps = [
        ...deps,
        ...Object.keys((json.devDependencies as T | undefined) || {}),
      ];
    }

    if (options.peerDependencies) {
      deps = [
        ...deps,
        ...Object.keys((json.peerDependencies as T | undefined) || {}),
      ];
    }

    return deps;
  } catch (e) {
    logger.warn('package.json is broken');
    return [];
  }
};

export const checkSwcHelpers = async (options: {
  appDirectory: string;
  externalHelpers: ExternalHelpers;
}) => {
  const { appDirectory, externalHelpers } = options;
  if (
    (typeof externalHelpers === 'object' &&
      externalHelpers.disableHelpersCheck) ||
    externalHelpers === false
  ) {
    return;
  }
  const deps = await getAllDeps(appDirectory, {
    dependencies: true,
    devDependencies: true,
  });
  const swcHelpersPkgName = '@swc/helpers';
  if (!deps.includes(swcHelpersPkgName)) {
    const local = await import('../locale');
    throw new Error(local.i18n.t(local.localeKeys.errors.externalHelpers));
  }
};
