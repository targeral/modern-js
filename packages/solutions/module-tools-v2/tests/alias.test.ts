import path from 'path';
import { fs } from '@modern-js/utils';
import { runCli, initBeforeTest } from './utils';
import {
  aliasBundleObjectDistPath,
  aliasBundlelessFunctionDistPath,
  aliasBundleFunctionDistPath,
  aliasBundlelessObjectDistPath,
} from './constants';

initBeforeTest();

beforeAll(() => {
  jest.mock('../src/utils/onExit.ts', () => {
    return {
      __esModule: true,
      addExitListener: jest.fn(() => 'mocked'),
    };
  });
});

describe('alias in js project', () => {
  const fixtureDir = path.join(__dirname, './fixtures/alias/js');
  it('object usage', async () => {
    const bundleConfigFile = path.join(fixtureDir, './bundle-object.config.ts');
    await runCli({
      argv: ['build'],
      configFile: bundleConfigFile,
      appDirectory: fixtureDir,
    });
    let distFilePath = path.join(
      fixtureDir,
      aliasBundleObjectDistPath,
      './index.js',
    );
    expect(fs.existsSync(distFilePath)).toBe(true);
    let content = fs.readFileSync(distFilePath, 'utf-8');
    expect(content.includes('hello world')).toBe(true);
    expect(content).toMatchSnapshot();

    const bundlelessConfigFile = path.join(
      fixtureDir,
      './bundleless-object.config.ts',
    );
    await runCli({
      argv: ['build'],
      configFile: bundlelessConfigFile,
      appDirectory: fixtureDir,
    });
    distFilePath = path.join(
      fixtureDir,
      aliasBundlelessObjectDistPath,
      './index.js',
    );
    expect(fs.existsSync(distFilePath)).toBe(true);

    content = fs.readFileSync(distFilePath, 'utf-8');
    expect(content.includes(`"./b"`)).toBe(true);
    expect(content).toMatchSnapshot();
  });

  it('function usage', async () => {
    let configFile = path.join(fixtureDir, './bundle-function.config.ts');
    await runCli({
      argv: ['build'],
      configFile,
      appDirectory: fixtureDir,
    });
    let distFilePath = path.join(
      fixtureDir,
      aliasBundleFunctionDistPath,
      './index.js',
    );
    expect(fs.existsSync(distFilePath)).toBe(true);

    let content = fs.readFileSync(distFilePath, 'utf-8');
    expect(content.includes('hello world')).toBe(true);
    expect(content).toMatchSnapshot();

    configFile = path.join(fixtureDir, './bundleless-function.config.ts');
    await runCli({
      argv: ['build'],
      configFile,
      appDirectory: fixtureDir,
    });
    distFilePath = path.join(
      fixtureDir,
      aliasBundlelessFunctionDistPath,
      './index.js',
    );
    expect(fs.existsSync(distFilePath)).toBe(true);

    content = fs.readFileSync(distFilePath, 'utf-8');
    expect(content.includes(`"./b"`)).toBe(true);
    expect(content).toMatchSnapshot();
  });
});

describe('alias in ts project', () => {
  const fixtureDir = path.join(__dirname, './fixtures/alias/ts');
  it('object usage', async () => {
    const bundleConfigFile = path.join(fixtureDir, './bundle-object.config.ts');
    await runCli({
      argv: ['build'],
      configFile: bundleConfigFile,
      appDirectory: fixtureDir,
    });
    let distFilePath = path.join(
      fixtureDir,
      aliasBundleObjectDistPath,
      './index.js',
    );
    expect(fs.existsSync(distFilePath)).toBe(true);

    let content = fs.readFileSync(distFilePath, 'utf-8');
    expect(content.includes('hello world')).toBe(true);
    expect(content).toMatchSnapshot();

    const bundlelessConfigFile = path.join(
      fixtureDir,
      './bundleless-object.config.ts',
    );
    await runCli({
      argv: ['build'],
      configFile: bundlelessConfigFile,
      appDirectory: fixtureDir,
    });
    distFilePath = path.join(
      fixtureDir,
      aliasBundlelessObjectDistPath,
      './index.js',
    );
    expect(fs.existsSync(distFilePath)).toBe(true);

    content = fs.readFileSync(distFilePath, 'utf-8');
    expect(content.includes(`"./b"`)).toBe(true);
    expect(content).toMatchSnapshot();
  });

  it('function usage', async () => {
    let configFile = path.join(fixtureDir, './bundle-function.config.ts');
    await runCli({
      argv: ['build'],
      configFile,
      appDirectory: fixtureDir,
    });
    let distFilePath = path.join(
      fixtureDir,
      aliasBundleFunctionDistPath,
      './index.js',
    );
    expect(fs.existsSync(distFilePath)).toBe(true);

    let content = fs.readFileSync(distFilePath, 'utf-8');
    expect(content.includes('hello world')).toBe(true);
    expect(content).toMatchSnapshot();

    configFile = path.join(fixtureDir, './bundleless-function.config.ts');
    await runCli({
      argv: ['build'],
      configFile,
      appDirectory: fixtureDir,
    });
    distFilePath = path.join(
      fixtureDir,
      aliasBundlelessFunctionDistPath,
      './index.js',
    );
    expect(fs.existsSync(distFilePath)).toBe(true);

    content = fs.readFileSync(distFilePath, 'utf-8');
    expect(content.includes(`"./b"`)).toBe(true);
    expect(content).toMatchSnapshot();
  });
});
