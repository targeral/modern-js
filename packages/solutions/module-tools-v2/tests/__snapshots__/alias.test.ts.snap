// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`alias in js project function usage 1`] = `
"var o=\\"hello world\\";console.info(o);
"
`;

exports[`alias in js project function usage 2`] = `
"import { b } from \\"./b\\";
console.info(b);
"
`;

exports[`alias in js project object usage 1`] = `
"var o=\\"hello world\\";console.info(o);
"
`;

exports[`alias in js project object usage 2`] = `
"import { b } from \\"./b\\";
console.info(b);
"
`;

exports[`alias in ts project function usage 1`] = `
"var o=\\"hello world\\";console.info(o);
"
`;

exports[`alias in ts project function usage 2`] = `
"import { b } from \\"./b\\";
console.info(b);
"
`;

exports[`alias in ts project object usage 1`] = `
"var o=\\"hello world\\";console.info(o);
"
`;

exports[`alias in ts project object usage 2`] = `
"import { b } from \\"./b\\";
console.info(b);
"
`;
